File: src/main/logger.ts
import { app } from 'electron'
import { existsSync, mkdirSync, createWriteStream, readdirSync, unlinkSync, type WriteStream } from 'fs'
import { join } from 'path'
import store from './store'

const LOG_PREFIX = 'anime-appimage-'
const MAX_LOG_FILES = 15

let stream: WriteStream | null = null
let logPath = ''

function pad(n: number): string {
  return String(n).padStart(2, '0')
}

function timestamp(): string {
  const d = new Date()
  return `${d.getFullYear()}-${pad(d.getMonth() + 1)}-${pad(d.getDate())} ${pad(d.getHours())}:${pad(d.getMinutes())}:${pad(d.getSeconds())}.${String(d.getMilliseconds()).padStart(3, '0')}`
}

function pruneOldLogs(dir: string): void {
  try {
    const files = readdirSync(dir)
      .filter((f) => f.startsWith(LOG_PREFIX) && f.endsWith('.log'))
      .sort()
    const excess = files.length - MAX_LOG_FILES
    for (let i = 0; i < excess; i++) {
      try {
        unlinkSync(join(dir, files[i]))
      } catch {
        // best-effort cleanup, ignore
      }
    }
  } catch {
    // dir listing failed, skip pruning
  }
}

function startStream(dir: string): string {
  pruneOldLogs(dir)
  const now = new Date()
  const stamp = `${now.getFullYear()}${pad(now.getMonth() + 1)}${pad(now.getDate())}-${pad(now.getHours())}${pad(now.getMinutes())}${pad(now.getSeconds())}`
  const unique = Math.random().toString(16).slice(2, 8)
  logPath = join(dir, `${LOG_PREFIX}${stamp}-${unique}.log`)
  stream = createWriteStream(logPath, { flags: 'a' })

  log('info', `=== Log iniciado: ${logPath} ===`)
  log(
    'info',
    `Plataforma: ${process.platform}, Electron: ${process.versions.electron}, App: ${app.getVersion()}`
  )
  log('info', `argv: ${process.argv.join(' ')}`)
  log(
    'info',
    `Sesión — XDG_SESSION_TYPE=${process.env.XDG_SESSION_TYPE ?? ''} WAYLAND_DISPLAY=${process.env.WAYLAND_DISPLAY ?? ''} DISPLAY=${process.env.DISPLAY ?? ''} SteamDeck=${process.env.SteamDeck ?? ''} SteamGamepadUI=${process.env.SteamGamepadUI ?? ''} SteamOS=${process.env.STEAM_RUNTIME ?? ''}`
  )
  return logPath
}

export function initLogger(): string {
  const libraryPath = store.get('libraryPath') || app.getPath('videos')
  const primaryDir = join(libraryPath, 'logs')
  try {
    if (!existsSync(primaryDir)) mkdirSync(primaryDir, { recursive: true })
    return startStream(primaryDir)
  } catch (err) {
    const fallbackDir = join(app.getPath('userData'), 'logs')
    if (!existsSync(fallbackDir)) mkdirSync(fallbackDir, { recursive: true })
    const path = startStream(fallbackDir)
    log('warn', `No se pudo escribir logs en ${primaryDir} (${(err as Error).message}), usando ${fallbackDir}`)
    return path
  }
}

export function log(level: 'info' | 'warn' | 'error', message: string): void {
  const line = `[${timestamp()}] [${level.toUpperCase()}] ${message}`
  if (level === 'error') console.error(line)
  else if (level === 'warn') console.warn(line)
  else console.log(line)
  stream?.write(line + '\n')
}

export function getLogPath(): string {
  return logPath
}

export function closeLogger(): void {
  stream?.end()
  stream = null
}

File: src-tauri/src/logger.rs
use chrono::Local;
use rand::Rng;
use std::fs::{self, File, OpenOptions};
use std::io::Write;
use std::path::PathBuf;
use std::sync::Mutex;

const LOG_PREFIX: &str = "anime-appimage-";
const MAX_LOG_FILES: usize = 15;

pub struct Logger {
    file: Mutex<Option<File>>,
}

impl Logger {
    pub fn init(library_path: Option<&str>, fallback_dir: PathBuf) -> Self {
        let primary = library_path.map(|p| PathBuf::from(p).join("logs"));

        let (dir, fallback_note) = match &primary {
            Some(dir) if fs::create_dir_all(dir).is_ok() => (dir.clone(), None),
            _ => {
                let _ = fs::create_dir_all(&fallback_dir);
                (fallback_dir, primary)
            }
        };

        prune_old_logs(&dir);

        let now = Local::now();
        let stamp = now.format("%Y%m%d-%H%M%S");
        let unique: u32 = rand::thread_rng().gen_range(0x1000..0xffffff);
        let log_path = dir.join(format!("{LOG_PREFIX}{stamp}-{unique:06x}.log"));

        let file = OpenOptions::new().create(true).append(true).open(&log_path).ok();
        let logger = Logger { file: Mutex::new(file) };

        logger.log("info", &format!("=== Log iniciado: {} ===", log_path.display()));
        logger.log(
            "info",
            &format!("Plataforma: {}, Tauri, App: {}", std::env::consts::OS, env!("CARGO_PKG_VERSION")),
        );
        let args: Vec<String> = std::env::args().collect();
        logger.log("info", &format!("argv: {}", args.join(" ")));
        logger.log(
            "info",
            &format!(
                "Sesión — XDG_SESSION_TYPE={} WAYLAND_DISPLAY={} DISPLAY={} SteamDeck={} SteamGamepadUI={}",
                std::env::var("XDG_SESSION_TYPE").unwrap_or_default(),
                std::env::var("WAYLAND_DISPLAY").unwrap_or_default(),
                std::env::var("DISPLAY").unwrap_or_default(),
                std::env::var("SteamDeck").unwrap_or_default(),
                std::env::var("SteamGamepadUI").unwrap_or_default(),
            ),
        );
        if let Some(fallback_from) = fallback_note {
            logger.log(
                "warn",
                &format!("No se pudo escribir logs en {}, usando carpeta de datos de la app", fallback_from.display()),
            );
        }

        logger
    }

    pub fn log(&self, level: &str, message: &str) {
        let line = format!(
            "[{}] [{}] {}",
            Local::now().format("%Y-%m-%d %H:%M:%S%.3f"),
            level.to_uppercase(),
            message
        );
        match level {
            "error" => eprintln!("{line}"),
            "warn" => eprintln!("{line}"),
            _ => println!("{line}"),
        }
        if let Some(file) = self.file.lock().unwrap().as_mut() {
            let _ = writeln!(file, "{line}");
        }
    }
}

fn prune_old_logs(dir: &PathBuf) {
    let Ok(entries) = fs::read_dir(dir) else { return };
    let mut files: Vec<PathBuf> = entries
        .filter_map(|e| e.ok())
        .map(|e| e.path())
        .filter(|p| {
            p.file_name()
                .and_then(|n| n.to_str())
                .map(|n| n.starts_with(LOG_PREFIX) && n.ends_with(".log"))
                .unwrap_or(false)
        })
        .collect();
    files.sort();
    if files.len() > MAX_LOG_FILES {
        for path in &files[..files.len() - MAX_LOG_FILES] {
            let _ = fs::remove_file(path);
        }
    }
}
